### ISSUE TITLE
can this be run with gmsa account

### ISSUE THREAD

**Issue by banchana** [Esri/arcgis-powershell-dsc#367]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

### Module Version

* 3.3.0

### Affected Resource(s)

* ArcGIS_Server
* ArcGIS_Portal

### Configuration Files

 "Version": "[10.8.1]",
        "ServerRole": "GeneralPurposeServer",
        "Credentials": {
            "ServiceAccount": {
                "Password": "[ServiceAccount Password]",
                "UserName": "[ServiceAccount Username - Can be a Domain Account]",
                "IsDomainAccount": true,
                "IsMSAAccount": false
            }
# Copy-paste your DSC JSON configurations here - for large configs,
# please use a service like Dropbox and share a link to the ZIP file.
```

### Expected Behavior

<!--- What should have happened? --->

### Actual Behavior

<!--- What actually happened? --->

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in a Azure China/Germany/Government? --->
can we use a gmsa (group managed service account), it does not have a password and currently it does not list that type of account as an option

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Such as vendor documentation?
--->

**Comment by cameronkroeker** [comment]
Hi @banchana,

Yes, the ArcGIS Module (v3.2.0 and above) does support using Group Managed Service Accounts (gMSA).

**Here is an example json:**

```json
"Credentials": {
     "ServiceAccount": {
           "UserName": "domain\\UserName$",
           "IsDomainAccount": false,
           "IsMSAAccount": true
  }
},
```

Thanks,
Cameron K.

**Comment by banchana** [comment]
Thanks @cameronkroeker . 
-In your example, you have set IsDomainAccount to false, but the group managed service account is also a domain account. 
-Why do we need to set IsMSAAcount to true? 
-I was using the version 3.3.0, so are you suggesting that i use the version 3.2.0 instead if i need to use group managed service account?

Can you please clarify, We are in the verge of using the traditional installer, but wanted to try this script out for the first time to see how this works for us.

**Comment by cameronkroeker** [comment]
> Thanks @cameronkroeker . -In your example, you have set IsDomainAccount to false, but the group managed service account is also a domain account. -Why do we need to set IsMSAAcount to true?
> 
> Can you please clarify, We are in the verge of using the traditional installer, but wanted to try this script out for the first time to see how this works for us.

Hi @banchana,

I would encourage you to use the latest version of the module, [v3.3.0](https://github.com/Esri/arcgis-powershell-dsc/releases/tag/v3.3.0).

Setting IsMSAAccount to true tells the module that the arcgis run-as account is a domain group managed service account, which will then ignore the Password attribute (since there isn't a password). You can set IsDomainAccount to true but this is isn't necessary and will be ignored. IsDomainAccount is meant for standard domain accounts that have a password.

Here are all the available attributes available that can be defined in your json config in v3.3.0:

https://github.com/Esri/arcgis-powershell-dsc/wiki/v3.3.0-Variables-reference-page-for-JSON-configuration-files

Thanks,
Cameron K.

**Comment by banchana** [comment]
Thanks @cameronkroeker . I will be working on this today then and will update, if i face any difficulties in the process.

**Comment by banchana** [comment]
I am trying to install an arcgis server with only the server role and with a web adaptor and would federate to an arcgis portal.
This is the command i used to invoke the install.

Invoke-ArcGISConfiguration -ConfigurationParametersFile [[Path to Configuration JSON File]] -Mode [Install] 

Here is what i got in my json config. I am getting errors. I if have my mode as InstallConfigureLicense, it would be nice to put an ecp number instead of the need to have the license file itself.
But i am receiving errors right at the invoke command, for parameter Mode Install, says does not belong to the set "Install, InstallLicense,........" I am sure i am missing something and not doing correctly. 


{
    "AllNodes": [
        {
            "NodeName": "[machine.domain.com]",
            "Role": ["Server"],
            "SslCertificates":{
		"Path":"X:\XX\cert.domain.com.pfx",
		"Password":"password",
		"CNameFQDN: "cert.domain.com",
		"Target": [Server],
		}
    ],
    "ConfigData": {
        "Version": "[10.8.1]",
        "Credentials": {
            "ServiceAccount": {                
                "UserName": "[domain\\account_gmsa$]",
                "IsDomainAccount": false
                "IsMSAAccount": true
            }
        },
      "ServerContext":"arcgis", 
        "Server": {
                "Installer": {
                "Path": "[\\filesharepath\10.8.1\ArcGISServer]",
                "InstallDir": "[C:\ArcGIS\Server]",
                "InstallDirPython": "[C:\\Python27] "
            },
            "ServerDirectoriesRootLocation": "C:\\arcgisserver\\directories",
            "ConfigStoreLocation": "C:\\arcgisserver\\config-store",
	    "ServerContext":"arcgis",
            "PrimarySiteAdmin": {
                "UserName": "[username]",
                "Password": "[password]"
            },     
	"WebAdaptor":{
		"Installer":{
			"Path":"[\\filesharepath]",			
			"AdminAccessEnabled": - [false],
			"WebSiteId": - 1
			}
		},      
        "Federation": {
            "PortalHostName": "[https://arcgisenterpriseportal.domain.com]",
            "PortalPort": "[7443]",
            "PortalContext": "[webadaptorinstancenameForArcGISPortal]",
            "PortalAdministrator": {
                "UserName": "username",
                "Password": "password"
            }
        }
    }
}

**Comment by cameronkroeker** [comment]
Hi @banchana,

The reason for having to use an authorization file (*.prvc or *.ecp) is because the module silently authorizes ArcGIS Server using the SoftwareAuthorizationWizard.exe which requires a file to be passed in and not an authorization number.

There are a few syntax errors in the json file, I did my best to clean it up, give this one a try:

```json
{
    "AllNodes": [
        {
            "NodeName": "machine.domain.com",
            "Role": [
                "Server"
            ],
            "SslCertificates":{
              "Path":"X:\\XX\\cert.domain.com.pfx",
              "Password":"password",
              "CNameFQDN": "cert.domain.com",
              "Target": [
                "Server"
              ]
            }
        }
    ],
    "ConfigData": {
       "Version": "10.8.1",
       "Credentials": {
            "ServiceAccount": {
                "UserName": "domain\\account_gmsa$",
                "IsDomainAccount": false,
                "IsMSAAccount": true
            }
       },
       "ServerContext": "arcgis",
        "Server": {
            "LicenseFilePath":  "\\\\filesharepath\\10.8.1\\server_auth_file.prvc",
            "Installer": {
                "Path": "\\\\filesharepath\\10.8.1\\ArcGISServer\\ArcGIS_Server_Windows_1081_175203.exe",
                "InstallDir": "C:\\ArcGIS\\Server",
                "InstallDirPython": "C:\\Python27"
             },
            "ServerDirectoriesRootLocation": "C:\\arcgisserver\\directories",
            "ConfigStoreLocation": "C:\\arcgisserver\\config-store",
            "ServerContext":"arcgis",
            "PrimarySiteAdmin": {
               "UserName": "username",
               "Password": "password"
            }
       },     
       "WebAdaptor":{
            "Installer":{
              "Path":"\\\\filesharepath\\ArcGIS_Web_Adaptor_for_Microsoft_IIS_1081_175217.exe"
            },
            "AdminAccessEnabled": false,
            "WebSiteId": 1
        },      
        "Federation": {
            "PortalHostName": "arcgisenterpriseportal.domain.com",
            "PortalPort": "443",
            "PortalContext": "webadaptorinstancenameForArcGISPortal",
            "PortalAdministrator": {
                "UserName": "username",
                "Password": "password"
            }
        }
    }
}
```

Thanks,
Cameron K.

**Comment by banchana** [comment]
Hi Cameron,
Thanks for reaching out.

No I had to just ditch this method. I kept getting some error and I needed
the installation be done by a deadline. It would be nice to get it to work
once.
I have the json file ready, if you want to come in to a screenshare
sometime after next week, I can do that. It will be easier that way. I will
be busy with esri dev conference next week, the digital one.

**Comment by cameronkroeker** [comment]
Hello @banchana,

I recommend submitting a request through Esri Technical Support as they will have the tools and resources available to setup a screen sharing session to assist.

https://support.esri.com/en/contact-tech-support

Thanks,
Cameron K.

**Comment by banchana** [comment]
Hi Cameron,
I am going to be working in standing up a new arcgis enterprise and would
be going through a lot of installation in a lot of servers. This could be
an opportunity for us to get honed in on the script based installations.

We will try and open ticket with support if we have any issues. I am going
to refer your name in the ticket, if that is okay with you.

Thanks
Banchana

**Comment by banchana** [comment]
Hi Cameron,
i have created a support case with esri Case #03061167.
This would be a really great opportunity for us to utilize this script and potentially for your team as well to see that it works successfully.

our deployment will be HA, so two portal server, two data store servers, two Hosting servers, one arcgis server site with three machine, one another arcgis server site with three machines.

Will this support that kind of a deployment?
I have some questions around the format of the json config mainly, was hoping to get some answers through the esri case.

Thanks
Banchana

**Comment by cameronkroeker** [comment]
Hi @banchana,

Thanks I will see if I can find out who is going to own the case and reach out to that person. In the mean time to address your json configuration question regarding multiple sites you will need to use different json files to accomplish this. I believe you will need at minimum of 3 json files (essentially a json file for each different AGS site):

- First json file will be for your Highly available base Enterprise deployment
  - 2 portals
  - 2 hosting servers
  - 2 data stores
- Second json file for your AGS Site that contains 3 nodes
- Third json file for your second AGS Site that contains 3 nodes

Thanks,
Cameron K.

**Comment by banchana** [comment]
okay, i will start crafting those json files then per your suggestion and will reach out through the case for any questions. thank you, we are really excited to make this work.

**Comment by cameronkroeker** [comment]
@banchana 

Here is an example json of a Highly available base Enterprise deployment (3 nodes - Primary, Standby, FileShare) with SSL certificates, web adaptors, and assuming you have an external load balancer. If you have your own file share pre-configured and do not want the module to create one then you can remove the third node that contains the "FileShare" role.

```json
{
    "AllNodes": [
        {
            "NodeName": "PrimaryNodeHostName",
            "DataStoreTypes": [
               "Relational",
               "TileCache"
            ],
            "Role": [
               "Server",
               "Portal", 
               "DataStore", 
               "ServerWebAdaptor",
               "PortalWebAdaptor"
            ],
            "SslCertificates": [{
                    "Path": "C:\\Certificates\\ca-certificate.pfx",
                    "CNameFQDN": "PrimaryNodeHostName.domain.com",
                    "Password": "password",
                    "Target": [
                       "Server",
                       "Portal",
                       "WebAdaptor",
                       "DataStore"
                    ]
                }
            ]
        },
        {
            "NodeName": "StandbyNodeHostName",
            "DataStoreTypes": [
               "Relational",
               "TileCache"
            ],
            "Role": [
               "Server",
               "Portal",
               "DataStore",
               "ServerWebAdaptor",
               "PortalWebAdaptor"
            ],
            "SslCertificates": [{
                    "Path": "C:\\Certificates\\ca-certificate.pfx",
                    "CNameFQDN": "StandbyNodeHostName.domain.com",
                    "Password": "password",
                    "Target": [
                       "Server", 
                       "Portal",
                       "WebAdaptor",
                       "DataStore"
                    ]
                }
            ]
        },
		{
            "NodeName": "FileShareNodeHostName",
            "Role": ["FileShare"]
        }
    ],
    "ConfigData": {
        "Version": "10.9.1",
        "FileShareLocalPath": "C:\\FileShare",
        "FileShareName": "FileShare",
        "ServerContext": "server",
        "PortalContext": "portal",
        "ServerRole": "GeneralPurposeServer",
        "Credentials": {
            "ServiceAccount": {
                "Password": "ServiceAccount Password",
                "UserName": "ServiceAccount Username - Can be a Domain Account",
                "IsDomainAccount": false,
				"IsMSAAccount": false
            }
        },
		"SslRootOrIntermediate": [
		    {
			     "Alias": "ca-certificate-root",
			     "Path": "C:\\Certificates\\ca-certificate-root.cer"
		    },
			{
			     "Alias": "ca-certificate-intermediate",
			     "Path": "C:\\Certificates\\ca-certificate-intermediate.cer"
		    }
	    ],
        "Portal": {
            "LicenseFilePath": "C:\\Authorization_Files\\portal1091.json",
            "PortalLicenseUserTypeId": "creatorUT",
            "Installer": {
                "Path": "C:\\1091_Final\\Portal_for_ArcGIS_Windows_1091_180052.exe",
                "InstallDir": "C:\\Program Files\\ArcGIS\\Portal",
                "ContentDir": "C:\\arcgisportal"
            },
            "PortalAdministrator": {
                "UserName": "portaladmin",
                "Email": "gisadmin@domain.com",
                "Password": "portaladmin",
                "SecurityQuestionIndex": 13,
                "SecurityAnswer": "Foo"
            },
            "ContentDirectoryLocation": "\\\\FileShareNodeHostName\\FileShare\\arcgisportal\\content",
            "EnableAutomaticAccountCreation": false,
			"DefaultRoleForUser": "account_viewer",
            "DefaultUserLicenseTypeIdForUser": "creatorUT",
            "ExternalLoadBalancer": "Ext-lb.domain.com"
        },
        "Server": {
            "LicenseFilePath": "C:\\Authorization_Files\\Version10.9\\server.prvc",
            "Installer": {
                "Path": "C:\\1091_Final\\ArcGIS_Server_Windows_1091_180041.exe",
                "InstallDir": "C:\\Program Files\\ArcGIS\\Server",
                "EnableArcMapRuntime": true,
                "EnableDotnetSupport": false
            },
            "ServerDirectoriesRootLocation": "\\\\FileShareNodeHostName\\FileShare\\arcgisserver\\directories",
            "ConfigStoreLocation": "\\\\FileShareNodeHostName\\FileShare\\arcgisserver\\config-store",
            "LocalRepositoryPath": "C:\\arcgisserver\\local",
            "ServerLogsLocation": "C:\\arcgisserver\\logs",
            "ServerDirectories": [{
                    "name": "arcgiscache",
                    "physicalPath": "\\\\FileShareNodeHostName\\FileShare\\arcgisserver\\directories\\arcgiscache",
                    "directoryType": "CACHE",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                }, {
                    "name": "arcgisjobs",
                    "physicalPath": "\\\\FileShareNodeHostName\\FileShare\\arcgisserver\\directories\\arcgisjobs",
                    "directoryType": "JOBS",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 360
                }, {
                    "name": "arcgisoutput",
                    "physicalPath": "\\\\FileShareNodeHostName\\FileShare\\arcgisserver\\directories\\arcgisoutput",
                    "directoryType": "OUTPUT",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 10
                }, {
                    "name": "arcgissystem",
                    "physicalPath": "\\\\FileShareNodeHostName\\FileShare\\arcgisserver\\directories\\arcgissystem",
                    "directoryType": "SYSTEM",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                }
            ],
            "PrimarySiteAdmin": {
                "UserName": "siteadmin",
                "Password": "siteadmin"
            },
            "ExternalLoadBalancer": "Ext-lb.domain.com"
        },
        "DataStore": {
            "ContentDirectoryLocation": "C:\\arcgisdatastore",
            "Installer": {
                "Path": "C:\\1091_Final\\ArcGIS_DataStore_Windows_1091_180054.exe",
                "InstallDir": "C:\\Program Files\\ArcGIS\\DataStore"
            },
            "EnableFailoverOnPrimaryStop": false
        },
        "WebAdaptor": {
            "AdminAccessEnabled": true,
            "Installer": {
                "Path": "C:\\1091_Final\\ArcGIS_Web_Adaptor_for_Microsoft_IIS_1091_180055.exe"
            }
        }
    }
}
```

Here is an example of a 3 node ArcGIS Server site, with a FileShare node and SSL certificates:

```json
{
    "AllNodes": [
        {
            "NodeName": "ServerMachineHostName1",
            "Role": [
                "Server",
                "ServerWebAdaptor"
            ],
            "SslCertificates": [{
                "Path": "C:\\Certificates\\ca-certificate.pfx",
                "CNameFQDN": "ServerMachineHostName1.domain.com",
                "Password": "password",
                "Target": [
                    "Server",
                    "WebAdaptor"
                ]
            }]
        },
        {
            "NodeName": "ServerMachineHostName2",
            "Role": [
                "Server",
                "ServerWebAdaptor"
            ],
            "SslCertificates": [{
                "Path": "C:\\Certificates\\ca-certificate.pfx",
                "CNameFQDN": "ServerMachineHostName2.domain.com",
                "Password": "password",
                "Target": [
                    "Server",
                    "WebAdaptor"
                ]
            }]
        },
        {
            "NodeName": "ServerMachineHostName3",
            "Role": [
                "Server",
                "ServerWebAdaptor"
            ],
            "SslCertificates": [{
                "Path": "C:\\Certificates\\ca-certificate.pfx",
                "CNameFQDN": "ServerMachineHostName3.domain.com",
                "Password": "password",
                "Target": [
                    "Server",
                    "WebAdaptor"
                ]
            }]
        },
        {
            "NodeName": "FileShareNodeHostName",
            "Role": [
                "FileShare"
            ]
        }
    ],
    "ConfigData": {
        "Version": "10.9.1",
        "ServerRole": "GeneralPurposeServer",
        "FileShareLocalPath": "C:\\AGSSiteFileShare",
        "FileShareName": "AGSSiteFileShare",
        "Credentials": {
            "ServiceAccount": {
                "Password": "[ServiceAccount Password]",
                "UserName": "[ServiceAccount Username - Can be a Domain Account]",
                "IsDomainAccount": false,
                "IsMSAAccount": false
            }
        },
        "Server": {
            "LicenseFilePath": "C:\\Authorization_Files\\Version10.9\\server.prvc",
            "Installer": {
                "Path": "C:\\1091_Final\\ArcGIS_Server_Windows_1091_180041.exe",
                "InstallDir": "C:\\Program Files\\ArcGIS\\Server",
                "EnableArcMapRuntime": true,
                "EnableDotnetSupport": false
            },
            "ServerDirectoriesRootLocation": "\\\\FileShareNodeHostName\\AGSSiteFileShare\\arcgisserver\\directories",
            "ConfigStoreLocation": "\\\\FileShareNodeHostName\\AGSSiteFileShare\\arcgisserver\\config-store",
            "LocalRepositoryPath": "C:\\arcgisserver\\local",
            "ServerLogsLocation": "C:\\arcgisserver\\logs",
            "ServerDirectories": [{
                    "name": "arcgiscache",
                    "physicalPath": "\\\\FileShareNodeHostName\\AGSSiteFileShare\\arcgisserver\\directories\\arcgiscache",
                    "directoryType": "CACHE",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                }, {
                    "name": "arcgisjobs",
                    "physicalPath": "\\\\FileShareNodeHostName\\AGSSiteFileShare\\arcgisserver\\directories\\arcgisjobs",
                    "directoryType": "JOBS",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 360
                }, {
                    "name": "arcgisoutput",
                    "physicalPath": "\\\\FileShareNodeHostName\\AGSSiteFileShare\\arcgisserver\\directories\\arcgisoutput",
                    "directoryType": "OUTPUT",
                    "cleanupMode": "TIME_ELAPSED_SINCE_LAST_MODIFIED",
                    "maxFileAge": 10
                }, {
                    "name": "arcgissystem",
                    "physicalPath": "\\\\FileShareNodeHostName\\AGSSiteFileShare\\arcgisserver\\directories\\arcgissystem",
                    "directoryType": "SYSTEM",
                    "cleanupMode": "NONE",
                    "maxFileAge": 0
                }
            ],
            "PrimarySiteAdmin": {
                "UserName": "siteadmin",
                "Password": "siteadmin"
            },
            "ExternalLoadBalancer": "Ext-lb.domain.com"
        },
        "Federation": {
            "PortalHostName": "[Portal Host Name (FQDN) with which this site needs to be federated]",
            "PortalPort": "[Port on which Portal is listening ex. - 443]",
            "PortalContext": "[Context of the Portal ex. - portal]",
            "PortalAdministrator": {
                "UserName": "admin",
                "Password": "changeit"
            }
        }
    }
}
```

For a full list of available attributes that can be used in your json configuration files check out:
https://github.com/Esri/arcgis-powershell-dsc/wiki/v3.3.2-Variables-reference-page-for-JSON-configuration-files

Hope this helps!

Thanks,
Cameron K.

**Comment by banchana** [comment]
Thanks Cameron! We have finished writing the json file for the base deployment with 2 web adaptor, 2 portal and 2 hosting. The example you sent helped a lot, so thank you so much and appreciate your help a lot.
I think i have some minor questions in each section, and have sent it over the esri support again. I have also attached the json file.
If we have this one configured, the other two for two sites should come along as well.
Hope you have a great weekend!
Thanks!
